### PR TITLE
[WIP] Nftables proxy

### PR DESCRIPTION
**Pull request by aojea**
/kind feature
```release-note
NONE
```

### CODE REVIEW

**Comment by aojea** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124272/pull-kubernetes-e2e-gce-100-performance/1778448759718416384

```
    {
      "data": {
        "Perc50": 1519.802112,
        "Perc90": 2282.213099,
        "Perc99": 6815.068537
      },
      "unit": "ms",
      "labels": {
        "Metric": "NetworkProgrammingLatency"
      }
```

This is much better, but iptables runs with 10s of sync period, last time we analyzed this the latency in iptables improved but CPU increased https://github.com/kubernetes/kubernetes/pull/110268#issuecomment-1326592664
https://github.com/kubernetes/kubernetes/pull/114229#issuecomment-1346566684
https://github.com/kubernetes/kubernetes/issues/115720

image

```
      "data": {
        "Perc50": 12.5,
        "Perc90": 22.5,
        "Perc99": 24.75
      },
      "unit": "ms",
      "labels": {
        "Metric": "InClusterNetworkLatency"
      }
```

is the same https://perf-dash.k8s.io/#/?jobname=gce-100Nodes-master&metriccategoryname=Network&metricname=Load_NetworkLatency&Metric=InClusterNetworkLatency


The sync rules in nftables

image


The sync rules in iptables from https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1778462542213943296 :) that are way better 

image

**Comment by danwinship** [comment]
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/124272/pull-kubernetes-e2e-gce-100-performance/1778448759718416384
> 
> ```
>     {
>       "data": {
>         "Perc50": 1519.802112,
>         "Perc90": 2282.213099,
>         "Perc99": 6815.068537
>       },
>       "unit": "ms",
>       "labels": {
>         "Metric": "NetworkProgrammingLatency"
>       }
> ```

(So that's the NetworkProgrammingLatency numbers from the `e2e-gce-100-performance` run you kicked off in this PR, aka "length of time from when the EndpointSliceController makes a change to when kube-proxy programs the change in nftables".)

> This is much better

(than equivalent current iptables). For comparision from a recent run:

```
    {
      "data": {
        "Perc50": 9034.440056,
        "Perc90": 10284.454043,
        "Perc99": 10984.283114
      },
      "unit": "ms",
      "labels": {
        "Metric": "NetworkProgrammingLatency"
      }
    }
```

> but iptables runs with 10s of sync period

(which is to say, the job runs with `--iptables-min-sync-period=10s` so that it does at most 1 iptables sync every 10 seconds, thus forcing p50 NetworkProgrammingLatency to hover around (10 seconds minus the length of time it takes to do a single sync).)

> last time we analyzed [switching to `--iptables-min-sync-period=1s`] the latency in iptables improved but CPU increased

Honestly, running kube-proxy with min sync period 10s is just crazy. Like, if a pod crashes, you want to keep sending new connections to its former IP address for another 10 seconds?

If we think kube-proxy uses too much CPU with a shorter sync period then we should fix that. (I think it's not clear how much of the CPU was kube-proxy itself vs kube-proxy plus iptables-restore?)

> ```
>       "data": {
>         "Perc50": 12.5,
>         "Perc90": 22.5,
>         "Perc99": 24.75
>       },
>       "unit": "ms",
>       "labels": {
>         "Metric": "InClusterNetworkLatency"
>       }
> ```
> 
> is the same

That's the one measuring packet latency for pod-to-service-IP-to-pod traffic?

How many services does this job create? We'd probably need at least 10s of thousands for the difference between iptables kube-proxy's rules and nftables kube-proxy's rules to become noticeable.

(Note that for purposes of this metric, you don't even need a large cluster to test that; you can create 1 pod and 10,000 services that all point to that same pod.)

> The sync rules in nftables
> ...
> The sync rules in iptables from https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-scalability/1778462542213943296 :) that are way better

So that's p95 of `sync_proxy_rules_duration_seconds`? (aka SyncProxyRulesLatency, aka the time it takes a single `syncProxyRules()` call to run)

The nftables proxy currently doesn't have the partial-syncing optimizations that iptables has (because we were waiting for the perf jobs before optimizing anything), so this is expected; every sync rewrites every rule, so nftables's `sync_proxy_rules_duration_seconds` ought to be more comparable to iptables's `sync_full_proxy_rules_duration_seconds` (note "`full`")

File: cluster/gce/config-test.sh [hunk 1]
```diff
@@ -532,8 +532,8 @@ KUBE_PROXY_DAEMONSET=${KUBE_PROXY_DAEMONSET:-false} # true, false
 # as an addon daemonset.
 KUBE_PROXY_DISABLE="${KUBE_PROXY_DISABLE:-false}" # true, false
 
-# Optional: Change the kube-proxy implementation. Choices are [iptables, ipvs].
-KUBE_PROXY_MODE=${KUBE_PROXY_MODE:-iptables}
+# Optional: Change the kube-proxy implementation. Choices are [iptables, ipvs, nftables].
+KUBE_PROXY_MODE=${KUBE_PROXY_MODE:-nftables}
```
[the remark this answers is not in the document]
Comment (aojea) [reply cluster/gce/config-test.sh:536]: it has to be removed , yes

**Comment by danwinship** [comment]
/close
replaced by https://github.com/kubernetes/kubernetes/pull/124383

**Pull request closed without merging**